kimchi/src/prover_index.rs

```diff
@@ -91,6 +100,45 @@ where
             max_poly_size,
             max_quot_size,
             fq_sponge_params,
+            verifier_index: None,
+            verifier_index_digest: None,
+        }
+    }
+
+    /// Retrieve or compute the digest for the corresponding verifier index.
+    /// If the digest is not already cached inside the index, store it.
+    pub fn compute_verifier_index_digest<
+        EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>,
+    >(
+        &mut self,
+    ) -> G::BaseField {
+        if let Some(verifier_index_digest) = self.verifier_index_digest {
+            return verifier_index_digest;
+        }
+
+        if let None = &self.verifier_index {
+            self.verifier_index = Some(self.verifier_index());
+        }
+
+        let verifier_index_digest = self.verifier_index_digest::<EFqSponge>();
+        self.verifier_index_digest = Some(verifier_index_digest);
+        verifier_index_digest
+    }
+
+    /// Retrieve or compute the digest for the corresponding verifier index.
+    pub fn verifier_index_digest<EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>>(
```

Comment:
this function should fill the cache if it's not already filled. This way it can be merged with `compute_verifier_index_digest`

Reply:
Same as all of the above: we need a version that doesn't mutate so that we don't require a mutable pointer to the index every time we want to create a proof.